docs/OperatorBestPractices.md

```diff
@@ -0,0 +1,103 @@
+
+* [Operator Providing Metadata in OLM](#operator-providing-metadata-in-olm)
+* [Operator Providing Metadata not in OLM](#operator-providing-metadata-not-in-olm)
+* [Operator Not Providing Metadata](#operator-not-providing-metadata)
+
+### Operator Providing Metadata in OLM
+
+This feature enables operator providers to specify binding information an
+operator's OLM (Operator Lifecycle Manager) descriptor. The Service Binding
+Operator extracts to bind the application together with the backing service.
+The information may be specified in the "status" and/or "spec" section of the
+OLM in plaintext or as a reference to a secret.
+
+For example, this is an example of the *bind-able* operator OLM Descriptor.
+In this case for a PostgreSQL database backing operator. 
+``` yaml
+---
+[...]
+statusDescriptors:
+  description: Name of the Secret to hold the DB user and password
+    displayName: DB Password Credentials
+    path: dbCredentials
+    x-descriptors:
+      - urn:alm:descriptor:io.kubernetes:Secret
+      - binding:env:object:secret:user
+      - binding:env:object:secret:password
+  description: Database connection IP address
+    displayName: DB IP address
+    path: dbConnectionIP
+    x-descriptors:
+      - binding:env:attribute
+```
+
+The operator is available [here.](https://github.com/operator-backing-service-samples/postgresql-operator)
+
+### Operator Providing Metadata not in OLM
+
+This feature enables operator prividers who do not use OLM (Operator Lifecycle
+Manager) to provide metadata outside of an OLM descriptor. In this method,
+the binding information is provided as annotations in the CRD of the operator
+that manages the backing service. The Service Binding Operator extracts the
+annotations to bind the application together with the backing service.
+
+For example: 
+``` yaml
+---
+[...]
+kind: CustomResourceDefinition
+apiVersion: apiextensions.k8s.io/v1beta1
+metadata:
+  name: databases.postgresql.baiju.dev
+  annotations:
+    servicebindingoperator.redhat.io/status.dbConfigMap.password: 'volume:configmap'
+    servicebindingoperator.redhat.io/status.dbConfigMap.username: 'env:configmap'
+    servicebindingoperator.redhat.io/status.dbName: env
+```
+
+### Operator Not Providing Metadata
+
+This feature enables operators that manage backing services but which don't
+have any metadata in their CSV to use the Service Binding Operator to bind
+together the service and applications. The Service Binding Operater binds all
+sub-resources defined in the backing service CR by populating the binding
+secret with information from Routes, Services, ConfigMaps, and Secrets owned
+by the backing service CR. The binding is initated by the introduction of
+this API option in the backing service CR:
+``` yaml
+detectBindingResources : true
+```
+When this API option is set to true, the Service Binding Operator 
+automatically detects Routes, Services, ConfigMaps, and Secrets owned by
+the backing service CR.
```

Comment:
How are potential naming collisions across these various backing service resources handled? How does the customEnvVar syntax distinguish between fields in each of the various backing service resources? Supporting three different binding methods as described in this document seems very confusing to me. I agree there is a need to support Operators that provide no metadata. Although it is a bit more complex and verbose, I think the proposal presented here: https://github.com/redhat-developer/service-binding-operator/issues/112#issuecomment-529271134 is more comprehensive and coherent than what is currently being described in this best practices guide.

Reply:
> Supporting three different binding methods as described in this document seems confusing and disjointed.

- CSV metadata will go away ( make it a bit OLM agnostic )
- CRD annotations will replace CSV metadata.
- CustomEnvVar is not fully baked now.

So, yes, there would be two ways to do things and I agree it is confusing. 

It needs more work and definition. I think https://github.com/redhat-developer/service-binding-operator/issues/112#issuecomment-529271134 is a alternative way to do `customEnvVar`. Since we are still alpha APIs, we will (un)happily :) break the API to remove `customEnvVar` after evaluating the alternatives - I can assure you that!

Reply:
OK. The annotations seem to include "env" or "volume" in their declaration. Don't you think the client developer rather than the CRD-developer should decide how these values are consumed? Inclusion of "env" or "volume" in the CRD couples extraction/selection and injection which seems undesirable.